File: Java/Lv : 4/징검 다리.java
import java.util.*;

public class Solution {
    
    public static boolean canAchieve(int minDist, int[] rocks, int n) {
        int deleteCount = 0;
        int prevRock = 0;
        
        for (int rock : rocks) {
            int dist = rock - prevRock;
            if (dist < minDist) {
                deleteCount++;
                if (deleteCount > n) {
                    return false;
                }
            } else {
                prevRock = rock;
            }
        }
        
        return true;
    }
    
    public static int solution(int distance, int[] rocks, int n) {
        int left = 1;
        int right = distance;
        
        Arrays.sort(rocks);
        int[] extendedRocks = new int[rocks.length + 1];
        System.arraycopy(rocks, 0, extendedRocks, 0, rocks.length);
        extendedRocks[rocks.length] = distance;
        
        int answer = 0;
        
        while (left <= right) {
            int mid = (left + right) / 2;
            if (canAchieve(mid, extendedRocks, n)) {
                answer = mid;
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        
        return answer;
    }
    
    public static void main(String[] args) {
        int distance = 25;
        int[] rocks = {2, 14, 11, 21, 17};
        int n = 2;
        int result = solution(distance, rocks, n);
        System.out.println(result);
    }
}

File: JavaScript/Lv : 4/징검 다리.js
// 이분 탐색 징검다리

function canAchieve(minDist, rocks, n) {
    let deleteCount = 0;
    let prevRock = 0;
    
    for (const rock of rocks) {
      const dist = rock - prevRock;
      if (dist < minDist) {
        deleteCount++;
        if (deleteCount > n) {
          return false;
        }
      } else {
        prevRock = rock;
      }
    }
    
    return true;
  }
  
  function solution(distance, rocks, n) {
    let left = 1;
    let right = distance;
    
    rocks.sort((a, b) => a - b);
    rocks.push(distance);
    
    let answer = 0;
    
    while (left <= right) {
      const mid = Math.floor((left + right) / 2);
      if (canAchieve(mid, rocks, n)) {
        answer = mid;
        left = mid + 1;
      } else {
        right = mid - 1;
      }
    }
    
    return answer;
  }
  